### ISSUE TITLE
How to control spaces between tags

### ISSUE THREAD

**Issue by infernalmaster** [jeromegn/slang#29]
So this
```
    a href="#"
      span
      | Contacts
```
is translated to such html:
```
<a href="#">
      <span>
      </span>
      Contacts
    </a>
```

But I want it to be without space between span and text like that:
```
<a href="#">
      <span>
      </span>Contacts
    </a>
```
Is it possible?

**Comment by jeromegn** [comment]
That's currently not implemented, the equivalent from slim would be: http://www.rubydoc.info/gems/slim/frames#Trailing_and_leading_whitespace_______

Something like:

```slim
    a href="#"
      span
      > Contacts
```

**Comment by infernalmaster** [comment]
Looks like slim trims all whitespaces between tags by default. So `>`, `<`, `'` for adding spaces not for deleting:  

> You can force Slim to add a trailing whitespace after a tag by adding a >.
> You can add a leading whitespace by adding <.

> The single quote tells Slim to copy the line (similar to |), but makes sure that a single trailing white space is appended.

So looks like that it's HAML removes spaces with that characters http://haml.info/docs/yardoc/file.REFERENCE.html#whitespace-removal--and-

I have time and need to help you with that but we should decide what we should to do with this

**Comment by jeromegn** [comment]
IIRC it’s either a rails thing or slim thing in production where it
compresses the HTML as much as possible.

Could be a setting with the parser.

**Comment by infernalmaster** [comment]
I found good workaround. I'm using ECR partials for that critical places.

**Comment by mig-hub** [comment]
Trimming all whitespace by default would be great.
It is also better for inline-block elements.

**Comment by robacarp** [comment]
Indeed, Slim does trim all whitespace by default. Haml injects it by default, like slang does, but haml also provides a way to prevent it from being added.

**Comment by robacarp** [comment]
I think this is a more obvious example. `<span>` tags are `display: inline`, so spacing around them matters more than around most html elements.

```slang
| un
span.red forget
| able
```

which renders

```
un
<span class="red">forget</span>
able
```

and therefore the rendered html looks like `un forget able`. 

The only way I've been able to get around this extra spacing is:

```slang
    == %(un<span class="red">forget</span>able)
```

**Comment by jeromegn** [comment]
I agree. I've had this problem before with inline elements with slim.

I'm wondering what's the best way to do this. Basically trim all whitespace everywhere between "nodes" (text and elements)? Unless text uses `'` (iirc, we don't support `<` or `>` yet.)

There are exceptions like special blocks for javascript or css.

**Comment by GloverDonovan** [comment]
How about using parentheses (or something similar) for inline Slang instead?

Example:

```slim
p This is un(span.red for(em get))able.
```

Output:

```html
<p>This is un<span class="red">for<em>get</em></span>able.</p>
```

**Comment by GloverDonovan** [comment]
How about using parentheses (or something similar) for inline Slang instead?

Input:

```slim
p Slang is a (span.color (em Slim)-inspired) templating language.
```

Output:

```html
<p>Slang is a <span class="color"><em>Slim</em>-inspired</span> templating language.</p>
```

**Comment by robacarp** [comment]
@jeromegn it's hard to say for sure, but given that HTML is _mostly_ ignorant of whitespace, it might be a safe first pass to just trim all whitespace. It would solve this problem, but I suspect it'd generate other problems.

Adding support for `>` and other whitespace manipulation, if possible, is probably a smart choice before making such a big change to the output every slang file generates. That would also bring Slang into closer parity with Slang, right?

**Comment by rperce** [comment]
@infernalmaster why did you close this? It seems like this is still a problem. 

FWIW, since this is mostly a problem with spans, for a personal workaround I've committed the cardinal sin of Parsing HTML With Regex and
```crystal
private macro view(filename, title argtitle = nil)
    title = {{argtitle}} || "{{filename.id}}"
    render("src/views/{{filename.id}}.slang", "src/views/Layout.slang")
        .gsub(/(<span[^>]+>)\n\s*/, "\\1") # newline after opening span
        .gsub(/<\/span>\n\s*<span/, "</span><span") # newline between spans
        .gsub(/\n\s*(<\/span>)/, "\\1") # newline before closing span
end
```
If there's a better way - if this *has* been solved properly and I just haven't noticed, or if I should be approaching the problem different, etc - I'd love to know!

**Comment by robacarp** [comment]
The reason this is more obvious with spans is that spans aren't block level elements. Added spacing into/around a block element is often not noticeable because it doesn't change flow in the document. For inline elements like span, link, strong, em, etc, the extra space matters more. For example, `<strong>this</strong>text` will show up as **this**text (no space) whereas `<div>this</div>text` will show up with a newline because the vanilla div styling forces it.

**Comment by infernalmaster** [comment]
Hey @rperce Apologies for the confusion, I closed it because I didn't see any activity in this thread, so I thought maybe it was fixed or no one cares

**Comment by rperce** [comment]
@robacarp yeah, totally understand *why* it's more relevant for spans, just wanted to share my workaround for anyone else running into this.

@infernalmaster reasonable! Thanks for the re-open - I would love to see this revisited.

**Comment by jeromegn** [comment]
Chiming in here: I don't think this is a _huge_ change.

I've been out of touch with Crystal, so I'd rather somebody else contribute a PR :)

I do think this is a useful feature!

**Comment by infernalmaster** [comment]
I did my best, but there are some limitations with embedded JS code. Some of problems present in inside master branch
